The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6671822851/job/18134525436) [(plain)](https://github.com/rust-lang/rust/commit/de98267eb87d43c8b54cfb98b7f5e6c2548a4f53/checks/18134525436/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=clubby789
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_fae99a4b-ae71-43cf-9505-51df360ded94
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=c8ac302933a8a8304c60bd4c06270cca48f8597c
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_fae99a4b-ae71-43cf-9505-51df360ded94
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_fae99a4b-ae71-43cf-9505-51df360ded94
GITHUB_TRIGGERING_ACTOR=clubby789
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117298/merge
GITHUB_WORKFLOW_SHA=c8ac302933a8a8304c60bd4c06270cca48f8597c
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---

9 LL |     pub struct Foo { text }
10    |         ++++++
11 
- error: expected one of `(` or `<`, found `{`
+   --> $DIR/recovered-block.rs:17:8
+    |
+    |
+ LL |     Foo { text: "".to_string() }
+ 
+ error: struct literal body without path
13   --> $DIR/recovered-block.rs:17:9
14    |
14    |
15 LL |     Foo { text: "".to_string() }

-    |         ^ expected one of `(` or `<`
+    |
+    |
+ help: you might have forgotten to add the struct literal inside the block
+    |
+ LL |     Foo { SomeStruct { text: "".to_string() } }
17 
- error: aborting due to 2 previous errors
+ error[E0308]: mismatched types
+   --> $DIR/recovered-block.rs:15:17
+   --> $DIR/recovered-block.rs:15:17
+    |
+ LL | pub fn bar() -> Foo {
+    |        ---      ^^^ expected `Foo`, found `()`
+    |        |
+    |        implicitly returns `()` as its body has no tail or `return` expression
+ error: aborting due to 4 previous errors
+ 
+ For more information about this error, try `rustc --explain E0308`.
20 
---
   |
LL |     pub Foo { text }
   |        ^
   |
help: add `struct` here to parse `Foo` as a public struct
LL |     pub struct Foo { text }
   |         ++++++

error: missing parameters for function definition
error: missing parameters for function definition
##[error]  --> /checkout/tests/ui/mismatched_types/recovered-block.rs:17:8
   |
LL |     Foo { text: "".to_string() }

error: struct literal body without path
##[error]  --> /checkout/tests/ui/mismatched_types/recovered-block.rs:17:9
   |
   |
LL |     Foo { text: "".to_string() }
   |
   |
help: you might have forgotten to add the struct literal inside the block
   |
LL |     Foo { SomeStruct { text: "".to_string() } }

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/mismatched_types/recovered-block.rs:15:17
   |
   |
LL | pub fn bar() -> Foo {
   |        ---      ^^^ expected `Foo`, found `()`
   |        |
   |        implicitly returns `()` as its body has no tail or `return` expression
error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0308`.
------------------------------------------
------------------------------------------


---- [ui] tests/ui/parser/issues/issue-105366.rs stdout ----
diff of stderr:

- error: you might have meant to write `impl` instead of `fn`
+ error: missing parameters for function definition
+   --> $DIR/issue-105366.rs:5:13
3    |
4 LL | fn From<i32> for Foo {
4 LL | fn From<i32> for Foo {
-    | ^^
+    |             ^ help: add a parameter list
+ 
+ error: expected one of `->`, `where`, or `{`, found keyword `for`
6    |
6    |
- help: replace `fn` with `impl` here
- LL | impl From<i32> for Foo {
-    | ~~~~
+ LL | fn From<i32> for Foo {
+ LL | fn From<i32> for Foo {
+    |              ^^^ expected one of `->`, `where`, or `{`
- error: aborting due to previous error
+ error: aborting due to 2 previous errors
13 
14 
---

2 
3 struct Foo;
4 
- impl From<i32> for Foo {
+ fn From<i32>() for Foo {
6     //~^ ERROR you might have meant to write `impl` instead of `fn`
7     fn from(_a: i32) -> Self {


The actual fixed differed from the expected fixed.
The actual fixed differed from the expected fixed.
Actual fixed saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/issues/issue-105366/issue-105366.fixed
To only update this specific test, also pass `--test-args parser/issues/issue-105366.rs`

error: 2 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/parser/issues/issue-105366.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/issues/issue-105366" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/issues/issue-105366/auxiliary"
--- stderr -------------------------------
error: missing parameters for function definition
##[error]  --> /checkout/tests/ui/parser/issues/issue-105366.rs:5:13
   |
   |
LL | fn From<i32> for Foo {
   |             ^ help: add a parameter list

error: expected one of `->`, `where`, or `{`, found keyword `for`
   |
LL | fn From<i32> for Foo {
LL | fn From<i32> for Foo {
   |              ^^^ expected one of `->`, `where`, or `{`
error: aborting due to 2 previous errors
------------------------------------------



---- [ui] tests/ui/parser/removed-syntax-fn-sigil.rs stdout ----
diff of stderr:

- error: expected `(`, found `~`
2   --> $DIR/removed-syntax-fn-sigil.rs:2:14
3    |
3    |
4 LL |     let x: fn~() = || ();
-    |         -    ^ expected `(`
-    |         |
-    |         while parsing the type for `x`
+    |              ^ help: add a parameter list
+    |              ^ help: add a parameter list
8 
- error: aborting due to previous error
+ error: expected one of `->`, `;`, or `=`, found `~`
+    |
+    |
+ LL |     let x: fn~() = || ();
+    |              ^ expected one of `->`, `;`, or `=`
+ error: aborting due to 2 previous errors
10 
11 

---
--- stderr -------------------------------
error: missing parameters for function definition
##[error]  --> /checkout/tests/ui/parser/removed-syntax-fn-sigil.rs:2:14
   |
LL |     let x: fn~() = || (); //~ ERROR expected `(`, found `~`


error: expected one of `->`, `;`, or `=`, found `~`
   |
   |
LL |     let x: fn~() = || (); //~ ERROR expected `(`, found `~`
   |              ^ expected one of `->`, `;`, or `=`
error: aborting due to 2 previous errors
------------------------------------------


```

</details>

--- a/compiler/rustc_parse/messages.ftl
+++ b/compiler/rustc_parse/messages.ftl
@@ -523,6 +523,9 @@ parse_missing_fn_for_function_definition = missing `fn` for function definition
 parse_missing_fn_for_method_definition = missing `fn` for method definition
     .suggestion = add `fn` here to parse `{$ident}` as a public method
 
+parse_missing_fn_params = missing parameters for function definition
+    .suggestion = add a parameter list
+
 parse_missing_for_in_trait_impl = missing `for` in a trait impl
     .suggestion = add `for` here
 

--- a/compiler/rustc_parse/src/errors.rs
+++ b/compiler/rustc_parse/src/errors.rs
@@ -1544,6 +1544,14 @@ pub(crate) enum AmbiguousMissingKwForItemSub {
 }
 
 #[derive(Diagnostic)]
+#[diag(parse_missing_fn_params)]
+pub(crate) struct MissingFnParams {
+    #[primary_span]
+    #[suggestion(code = "()", applicability = "machine-applicable", style = "short")]
+    pub span: Span,
+}
+
+#[derive(Diagnostic)]
 #[diag(parse_missing_trait_in_trait_impl)]
 pub(crate) struct MissingTraitInTraitImpl {
     #[primary_span]

--- a/compiler/rustc_parse/src/parser/item.rs
+++ b/compiler/rustc_parse/src/parser/item.rs
@@ -2492,6 +2492,16 @@ impl<'a> Parser<'a> {
     pub(super) fn parse_fn_params(&mut self, req_name: ReqName) -> PResult<'a, ThinVec<Param>> {
         let mut first_param = true;
         // Parse the arguments, starting out with `self` being allowed...
+        if self.token.kind != TokenKind::OpenDelim(Delimiter::Parenthesis)
+        // might be typo'd trait impl, handled elsewhere
+        && !self.token.is_keyword(kw::For)
+        {
+            // recover from missing argument list, e.g. `fn main -> () {}`
+            self.sess
+                .emit_err(errors::MissingFnParams { span: self.prev_token.span.shrink_to_hi() });
+            return Ok(ThinVec::new());
+        }
+
         let (mut params, _) = self.parse_paren_comma_seq(|p| {
             p.recover_diff_marker();
             let param = p.parse_param_general(req_name, first_param).or_else(|mut e| {

--- a/tests/ui/mismatched_types/recovered-block.rs
+++ b/tests/ui/mismatched_types/recovered-block.rs
@@ -12,10 +12,4 @@ pub fn foo() -> Foo {
 }
 //~^^ ERROR missing `struct` for struct definition
 
-pub fn bar() -> Foo {
-    fn
-    Foo { text: "".to_string() }
-}
-//~^^ ERROR expected one of `(` or `<`, found `{`
-
 fn main() {}

--- a/tests/ui/mismatched_types/recovered-block.stderr
+++ b/tests/ui/mismatched_types/recovered-block.stderr
@@ -9,11 +9,5 @@ help: add `struct` here to parse `Foo` as a public struct
 LL |     pub struct Foo { text }
    |         ++++++
 
-error: expected one of `(` or `<`, found `{`
-  --> $DIR/recovered-block.rs:17:9
-   |
-LL |     Foo { text: "".to_string() }
-   |         ^ expected one of `(` or `<`
-
-error: aborting due to 2 previous errors
+error: aborting due to previous error
 

--- a/tests/ui/parser/issues/issue-108109-fn-missing-params.fixed
+++ b/tests/ui/parser/issues/issue-108109-fn-missing-params.fixed
@@ -1,0 +1,9 @@
+// run-rustfix
+
+pub fn missing() -> () {}
+//~^ ERROR missing parameters for function definition
+
+pub fn missing2() {}
+//~^ ERROR missing parameters for function definition
+
+fn main() {}

--- a/tests/ui/parser/issues/issue-108109-fn-missing-params.rs
+++ b/tests/ui/parser/issues/issue-108109-fn-missing-params.rs
@@ -1,0 +1,9 @@
+// run-rustfix
+
+pub fn missing -> () {}
+//~^ ERROR missing parameters for function definition
+
+pub fn missing2 {}
+//~^ ERROR missing parameters for function definition
+
+fn main() {}

--- a/tests/ui/parser/issues/issue-108109-fn-missing-params.stderr
+++ b/tests/ui/parser/issues/issue-108109-fn-missing-params.stderr
@@ -1,0 +1,14 @@
+error: missing parameters for function definition
+  --> $DIR/issue-108109-fn-missing-params.rs:3:15
+   |
+LL | pub fn missing -> () {}
+   |               ^ help: add a parameter list
+
+error: missing parameters for function definition
+  --> $DIR/issue-108109-fn-missing-params.rs:6:16
+   |
+LL | pub fn missing2 {}
+   |                ^ help: add a parameter list
+
+error: aborting due to 2 previous errors
+

--- a/tests/ui/parser/removed-syntax-fn-sigil.rs
+++ b/tests/ui/parser/removed-syntax-fn-sigil.rs
@@ -1,3 +1,4 @@
 fn main() {
-    let x: fn~() = || (); //~ ERROR expected `(`, found `~`
+    let x: fn~() = || (); //~ ERROR missing parameters for function definition
+    //~| ERROR expected one of `->`, `;`, or `=`, found `~`
 }

--- a/tests/ui/parser/removed-syntax-fn-sigil.stderr
+++ b/tests/ui/parser/removed-syntax-fn-sigil.stderr
@@ -1,10 +1,14 @@
-error: expected `(`, found `~`
+error: missing parameters for function definition
   --> $DIR/removed-syntax-fn-sigil.rs:2:14
    |
 LL |     let x: fn~() = || ();
-   |         -    ^ expected `(`
-   |         |
-   |         while parsing the type for `x`
+   |              ^ help: add a parameter list
 
-error: aborting due to previous error
+error: expected one of `->`, `;`, or `=`, found `~`
+  --> $DIR/removed-syntax-fn-sigil.rs:2:14
+   |
+LL |     let x: fn~() = || ();
+   |              ^ expected one of `->`, `;`, or `=`
+
+error: aborting due to 2 previous errors
 

